tests/utils_tests/test_choices.py

```diff
@@ -0,0 +1,216 @@
+            match medium:
+                case "audio":
+                    yield from (("vinyl", _("Vinyl")), ("cd", _("CD")))
+                case "video":
+                    yield from (("vhs", _("VHS Tape")), ("dvd", _("DVD")))
+                case None:
+                    yield "Audio", generator(medium="audio")
+                    yield "Video", generator(medium="video")
+                    yield "unknown", _("Unknown")
+
+        choices = generator()
+        self.assertEqual(normalize_field_choices(choices), self.expected_nested)
+
+    def test_iterable_non_canonical(self):
+        # Canonical form is list of 2-tuple, but ensure lists all the way down work.
+        choices = [
+            ["C", _("Club")],
+            ["D", _("Diamond")],
+            ["H", _("Heart")],
+            ["S", _("Spade")],
+        ]
+        self.assertEqual(normalize_field_choices(choices), self.expected)
+
+    def test_iterator_non_canonical(self):
+        # Canonical form is list of 2-tuple, but ensure lists all the way down work.
+        def generator():
+            yield ["C", _("Club")]
+            yield ["D", _("Diamond")]
+            yield ["H", _("Heart")]
+            yield ["S", _("Spade")]
+
+        choices = generator()
+        self.assertEqual(normalize_field_choices(choices), self.expected)
+
+    def test_nested_iterable_non_canonical(self):
+        # Canonical form is list of 2-tuple, but ensure lists all the way down work.
+        choices = [
+            ["Audio", [["vinyl", _("Vinyl")], ["cd", _("CD")]]],
+            ["Video", [["vhs", _("VHS Tape")], ["dvd", _("DVD")]]],
+            ["unknown", _("Unknown")],
+        ]
+        self.assertEqual(normalize_field_choices(choices), self.expected_nested)
+
+    def test_nested_iterator_non_canonical(self):
+        # Canonical form is list of 2-tuple, but ensure lists all the way down work.
+        def generator():
+            yield ["Audio", [["vinyl", _("Vinyl")], ["cd", _("CD")]]]
+            yield ["Video", [["vhs", _("VHS Tape")], ["dvd", _("DVD")]]]
+            yield ["unknown", _("Unknown")]
+
+        choices = generator()
+        self.assertEqual(normalize_field_choices(choices), self.expected_nested)
+
+    def test_nested_mixed_mapping_and_iterable(self):
+        # Although not documented as it's better to stick to either mappings or
+        # iterables, ensure that this works as expected.
+        choices = {
+            "Audio": [("vinyl", _("Vinyl")), ("cd", _("CD"))],
+            "Video": [("vhs", _("VHS Tape")), ("dvd", _("DVD"))],
+            "unknown": _("Unknown"),
+        }
+        self.assertEqual(normalize_field_choices(choices), self.expected_nested)
+        choices = [
+            ("Audio", {"vinyl": _("Vinyl"), "cd": _("CD")}),
+            ("Video", {"vhs": _("VHS Tape"), "dvd": _("DVD")}),
+            ("unknown", _("Unknown")),
+        ]
+        self.assertEqual(normalize_field_choices(choices), self.expected_nested)
```

Comment:
Although not documented, as not recommended, this is going to inevitably happen. Earlier versions of this patch supported mappings at the top level with lists of 2-tuples for nested items, but not the other way around. This was resolve when I made `normalize_field_choices()` recursive and should hopefully avoid any "invalid" bug reports down the line. Besides, it's more consistent.

Reply:
There is one there at the top? 😕 I'll tweak it a little bit though...